packages/component/src/SendBox/TextBox.js

```diff
@@ -97,34 +98,158 @@ function useTextBoxSubmit() {
+        forwardedRef(element);
+      } else if (forwardedRef) {
+        forwardedRef.current = element;
+      }
+
+      inputElementRef.current = element;
+    },
+    [forwardedRef, inputElementRef]
+  );
 
   const sendBoxString = localize('TEXT_INPUT_ALT');
   const typeYourMessageString = localize('TEXT_INPUT_PLACEHOLDER');
 
-  const handleChange = useCallback(({ target: { value } }) => setTextBoxValue(value), [setTextBoxValue]);
+  const rememberInputState = useCallback(() => {
+    const {
+      current: { selectionEnd, selectionStart, value }
+    } = inputElementRef;
+
+    prevInputStateRef.current = { selectionEnd, selectionStart, value };
+  }, [inputElementRef, prevInputStateRef]);
+
+  // This is for TypeFocusSink. When the focus in on the script, then starting press "a", without this line, it would cause errors.
+  // We call rememberInputState() when "onFocus" event is fired, but since this is from TypeFocusSink, we are not able to receive "onFocus" event before it happen.
+  useEffect(rememberInputState, [rememberInputState]);
+
+  // This is for moving the selection while setting the send box value.
+  // If we only use setSendBox, we will need to wait for the next render cycle to get the value in, before we can set selectionEnd/Start.
+  const setSelectionRangeAndValue = useCallback(
+    ({ selectionEnd, selectionStart, value }) => {
+      if (inputElementRef.current) {
+        // We need to set the value, before selectionStart/selectionEnd.
+        inputElementRef.current.value = value;
+
+        inputElementRef.current.selectionStart = selectionStart;
+        inputElementRef.current.selectionEnd = selectionEnd;
+      }
+
+      setSendBox(value);
+    },
+    [inputElementRef, setSendBox]
+  );
+
+  const handleChange = useCallback(
+    event => {
+      const {
+        target: { selectionEnd, selectionStart, value }
+      } = event;
+
+      if (placeCheckpointOnChangeRef.current) {
+        undoStackRef.current.push({ ...prevInputStateRef.current });
+
+        placeCheckpointOnChangeRef.current = false;
+      }
+
+      const nextInputState = setTextBoxValue(value, { selectionEnd, selectionStart });
+
+      // If an emoticon is converted to emoji, place another checkpoint.
+      if (nextInputState.value !== value) {
+        undoStackRef.current.push({ selectionEnd, selectionStart, value });
+
+        placeCheckpointOnChangeRef.current = true;
+
+        setSelectionRangeAndValue(nextInputState);
+      }
+    },
+    [placeCheckpointOnChangeRef, prevInputStateRef, setSelectionRangeAndValue, setTextBoxValue, undoStackRef]
+  );
+
+  const handleFocus = useCallback(() => {
+    rememberInputState();
+
+    placeCheckpointOnChangeRef.current = true;
+  }, [placeCheckpointOnChangeRef, rememberInputState]);
+
+  const handleKeyDown = useCallback(
+    event => {
+      const { ctrlKey, key, metaKey } = event;
+
+      if ((ctrlKey || metaKey) && (key === 'Z' || key === 'z')) {
```

Comment:
Is `ctrl+Z` typical for undo?

Reply:
<kbd>CTRL</kbd> + <kbd>Z</kbd> is de facto on Windows. And I guess <kbd>COMMAND</kbd> + <kbd>Z</kbd> is de facto on Mac.

You spotted the corner I cut here. 😁

Please let me know if I should only allow <kbd>COMMAND</kbd> + <kbd>Z</kbd> on Mac but not <kbd>CTRL</kbd> + <kbd>Z</kbd>.